apps/pipeline-visualization/src/mermaid_builder.ts

```diff
@@ -0,0 +1,88 @@
+// SPDX-FileCopyrightText: 2023 Friedrich-Alexander-Universitat Erlangen-Nurnberg
+//
+// SPDX-License-Identifier: AGPL-3.0-only
+
+import { BlockDefinition } from '@jvalue/jayvee-language-server';
+
+export class MermaidBuilder {
+  protected mermaidLines: string[] = [];
+
+  pipelineHead(pipelineName: string): MermaidBuilder {
+    this.mermaidLines.push('subgraph ' + pipelineName);
+    return this;
+  }
+
+  blockType(block: BlockDefinition): MermaidBuilder {
+    this.mermaidLines.push(
+      `${block.name}[${block.name}<br><i>${block.type.ref?.name}</i>]`,
+    );
+    return this;
+  }
+
+  blockTypeWithProperties(
+    block: BlockDefinition,
+    propertyString: string,
+  ): MermaidBuilder {
+    this.mermaidLines.push(
+      `${block.name}[${block.name}<br><i>${block.type.ref?.name}</i><br>${propertyString}]`,
+    );
+    return this;
+  }
+
+  branching(block: BlockDefinition, child: BlockDefinition): MermaidBuilder {
+    this.mermaidLines.push(`${block.name}-->${child.name}`);
+    return this;
+  }
+
+  chain(chain: string[]): MermaidBuilder {
+    this.mermaidLines.push(chain.join('-->'));
+    return this;
+  }
+
+  compositeHeader(
+    block: BlockDefinition,
+    depth: number,
+    subgraphDirection: string,
+  ): MermaidBuilder {
+    this.mermaidLines.push(
+      `${'\t'.repeat(depth)}subgraph ${block.name}\n ${'\t'.repeat(
+        depth,
+      )}direction ${subgraphDirection}`,
+    );
+    return this;
+  }
+
+  compositeBody(subblocks: string[], depth: number): MermaidBuilder {
```

Comment:
`depth` can be an internal variable. It's the builder's job to make the output nice.
Opening a composite then increases the depth, closing decreases it.

A more common name for _depth_ in this context is _indent_ btw :)